### PR TITLE
Fix CPU Inductor FX cache thread specialization

### LINKED ISSUE #110611: torch.compile CPU backend is slower than eager for several transcendental functions
I tried a handful of somewhat important transcendentals (sigmoid, tanh, log, logit), and they all appear to be slower than eager.  On my 96c/192t Intel(R) Xeon(R) Platinum 8339HC CPU:

```
$ OMP_NUM_THREADS=1 numactl -C 94 python bench.py
log
eager:    363.3282147347927
compiled: 1091.864187270403
sigmoid
eager:    756.7794565111399
compiled: 872.3638635128736
logit
eager:    731.156948953867
compiled: 1594.5508629083633
tanh
eager:    475.9031366556883
compiled: 867.8419776260853

$ python bench.py
log
eager:    28.686294332146645
compiled: 106.62426799535751
sigmoid
eager:    91.70412458479404
compiled: 108.63539576530457
logit
eager:    25.92829428613186
compiled: 88.3821751922369
tanh
eager:    28.41925248503685
compiled: 113.94776217639446
```

```
import time

import torch

def do_bench(fn, iters=1000):
    fn()  # warmup
    s = time.perf_counter()
    for _ in range(iters):
        fn()
    e = time.perf_counter()

    return (e - s) /  iters * 1e6

x = torch.clamp(torch.rand(1 << 20), 1e-6, 1 - 1e-6)  # good range for log, logit

for fn in [torch.log,
           torch.sigmoid,
           torch.logit,
           torch.tanh,
           ]:
    cfn = torch.compile(fn)
    print(fn.__name__)

    eager = do_bench(lambda: fn(x))
    compiled = do_bench(lambda: cfn(x))

    print(f"eager:    {eager}")
    print(f"compiled: {compiled}")
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### LINKED ISSUE #167463: Number of threads is not part of guard
### 🐛 Describe the bug

Related to #167459 , the number of threads is hardcoded in compile cache, and recompilation does not happne for the same function but different number of threads. Obviously it causes performance issue. If setting dynamic number of threads hurt performance, one should use number of threads as part of guards.

### Error logs

_No response_

### Versions

Collecting environment information...
PyTorch version: 2.9.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: 9.0.0 (https://github.com/conda-forge/clangdev-feedstock 284a3d5d88509307bcfba64b055653ee347371db)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.11.11 (main, Dec 11 2024, 16:28:39) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.14.0-33-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           39 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  8
On-line CPU(s) list:                     0-7
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Core(TM) i7-9700K CPU @ 3.60GHz
CPU family:                              6
Model:                                   158
Thread(s) per core:                      1
Core(s) per socket:                      8
Socket(s):                               1
Stepping:                                13
CPU(s) scaling MHz:                      93%
CPU max MHz:                             4900.0000
CPU min MHz:                             800.0000
BogoMIPS:                                7200.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust sgx bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp vnmi md_clear flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               256 KiB (8 instances)
L1i cache:                               256 KiB (8 instances)
L2 cache:                                2 MiB (8 instances)
L3 cache:                                12 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-7
Vulnerability Gather data sampling:      Mitigation; Microcode (locked)
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             KVM: Mitigation: Split huge pages
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT disabled
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Mitigation; Microcode
Vulnerability Tsx async abort:           Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] intel-cmplr-lib-ur==2025.2.1
[pip3] intel-openmp==2025.2.1
[pip3] mkl==2025.1.0
[pip3] mkl-service==2.5.2
[pip3] numpy==2.2.0
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] tbb==2022.2.0
[pip3] tcmlib==1.4.0
[pip3] torch==2.9.0
[pip3] torch-xla==2.8.1
[pip3] torchviz==0.0.3
[pip3] triton==3.5.0
[pip3] umf==0.11.0
[conda] cuda-cudart               12.9.37                       0    nvidia
[conda] cuda-cudart-dev           12.9.37                       0    nvidia
[conda] cuda-cudart-dev_linux-64  12.9.37                       0    nvidia
[conda] cuda-cudart-static        12.9.37                       0    nvidia
[conda] cuda-cudart-static_linux-64 12.9.37                       0    nvidia
[conda] cuda-cudart_linux-64      12.9.37                       0    nvidia
[conda] cuda-cupti                12.9.19                       0    nvidia
[conda] cuda-cupti-dev            12.9.19                       0    nvidia
[conda] cuda-libraries            12.9.0                        0    nvidia
[conda] cuda-libraries-dev        12.9.0                        0    nvidia
[conda] cuda-libraries-static     12.9.0                        0    nvidia
[conda] cuda-nvrtc                12.9.41                       0    nvidia
[conda] cuda-nvrtc-dev            12.9.41                       0    nvidia
[conda] cuda-nvrtc-static         12.9.41                       0    nvidia
[conda] cuda-nvtx                 12.9.19                       0    nvidia
[conda] cuda-opencl               12.9.19                       0    nvidia
[conda] cuda-opencl-dev           12.9.19                       0    nvidia
[conda] intel-cmplr-lib-ur        2025.2.1                 pypi_0    pypi
[conda] intel-openmp              2025.2.1                 pypi_0    pypi
[conda] libcublas                 12.9.0.13                     0    nvidia
[conda] libcublas-dev             12.9.0.13                     0    nvidia
[conda] libcublas-static          12.9.0.13                     0    nvidia
[conda] libcufft                  11.4.0.6                      0    nvidia
[conda] libcufft-dev              11.4.0.6                      0    nvidia
[conda] libcufft-static           11.4.0.6                      0    nvidia
[conda] libcurand                 10.3.10.19                    0    nvidia
[conda] libcurand-dev             10.3.10.19                    0    nvidia
[conda] libcurand-static          10.3.10.19                    0    nvidia
[conda] libcusolver               11.7.4.40                     0    nvidia
[conda] libcusolver-dev           11.7.4.40                     0    nvidia
[conda] libcusolver-static        11.7.4.40                     0    nvidia
[conda] libcusparse               12.5.9.5                      0    nvidia
[conda] libcusparse-dev           12.5.9.5                      0    nvidia
[conda] libcusparse-static        12.5.9.5                      0    nvidia
[conda] libnvjitlink              12.9.41                       0    nvidia
[conda] libnvjitlink-dev          12.9.41                       0    nvidia
[conda] libnvjitlink-static       12.9.41                       0    nvidia
[conda] mkl                       2025.1.0                 pypi_0    pypi
[conda] mkl-service               2.5.2                    pypi_0    pypi
[conda] nomkl                     3.0                           0  
[conda] numpy                     2.2.0                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.8.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.8.90                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.8.93                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.8.90                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.10.2.21                pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.3.83                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.9.90                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.3.90                pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.8.93                pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.7.1                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.27.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.8.93                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.8.90                  pypi_0    pypi
[conda] tbb                       2022.2.0                 pypi_0    pypi
[conda] tcmlib                    1.4.0                    pypi_0    pypi
[conda] torch                     2.9.0                    pypi_0    pypi
[conda] torch-xla                 2.8.1                    pypi_0    pypi
[conda] torchviz                  0.0.3                    pypi_0    pypi
[conda] triton                    3.5.0                    pypi_0    pypi
[conda] umf                       0.11.0                   pypi_0    pypi

cc @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184444

Include the resolved CPU runtime thread count in the FX graph cache key when C++ codegen may specialize on it, including no-input CPU factory graphs, so cached serial kernels are not reused under a different thread count. Making the resolved thread count part of the FX graph cache key fixes all reported cases where a CPU graph compiled under one `torch.set_num_threads()` value was incorrectly served from cache after the thread count changed.

Fixes #110611
Fixes #167463
Fixes #160812

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by oulgen — APPROVED** [review]

**Comment by jansel** [comment]
Agent, share some benchmarking results to prove this fixes the perf issues mentioned in the issue.

**Comment by jansel** [comment]
My agent says I reran the issue benchmark on PR head 56d1716c09e048407dbdbf6347e8a7f62f401f50 with a shared TORCHINDUCTOR_CACHE_DIR: the first subprocess populated the cache with OMP_NUM_THREADS=1/MKL_NUM_THREADS=1, then the second subprocess ran with the default 192 CPU threads. The default-thread subprocess correctly missed the FX graph cache (`hit=0 miss=4`) and stayed close to eager instead of reusing one-thread kernels:

| function | eager us/iter | compiled us/iter |
| --- | ---: | ---: |
| log | 72.209 | 91.022 |
| sigmoid | 86.872 | 95.630 |
| logit | 75.862 | 97.014 |
| tanh | 88.687 | 93.699 |

For comparison, the pre-fix reproduction of this same one-thread-then-default cache sequence had default-thread compiled times around 582/433/844/1543 us for log/sigmoid/logit/tanh, so this removes the severe cache-reuse slowdown path.

**Comment by jansel** [comment]
My agent says I rebased the stack onto current main after the mergebot run failed on ROCm trunk jobs, then resubmitted the PR as head 45fdc6786170d63ef9b17796effba20678ddf472. Validation passed with `PYTHONPATH=. python test/inductor/test_codecache.py -k test_hash_kwargs`, `PYTHONPATH=. python test/inductor/test_codecache.py -k test_cpu_thread_count`, and `lintrunner -a`.

**Comment by jansel** [comment]
My agent says I inspected the current Dr. CI new failures on head 45fdc6786170d63ef9b17796effba20678ddf472. The failures all happened during checkout/setup before the relevant tests or lints ran, with GitHub fetch/submodule/promisor credential errors and missing workflow scripts after incomplete checkout. I reran the failed Lint, pull, and inductor workflow jobs; CI is pending again.

**Comment by jansel** [comment]
My agent says I rebased the stack onto current main after the latest mergebot trunk build failure and resubmitted PR #184444 as head 96ea7fc04eac3aae2cc122fe16a37867f565cc4a. Validation passed with `PYTHONPATH=. python test/inductor/test_codecache.py -k test_hash_kwargs`, `PYTHONPATH=. python test/inductor/test_codecache.py -k test_cpu_thread_count`, `lintrunner -a`, and an independent review found no issues.

**Comment by jansel** [comment]
My agent says I inspected the current Dr. CI new failures on head 96ea7fc04eac3aae2cc122fe16a37867f565cc4a. The ROCm distributed failure is in `test/distributed/test_aten_comm_compute_reordering.py::TestComputeCommReorderingMultiProc::test_sink_waits`, and the macOS MPS job died with exit code 139; neither points at the CPU Inductor FX cache-key change. I requested reruns, and GitHub has the macOS MPS rerun in progress now; the ROCm job remains failed and cannot be rerun again until the current workflow attempt finishes, so this needs a CI revisit.

**Comment by jansel** [comment]
My agent says the macOS MPS rerun has passed. I reran the remaining Dr. CI new failure, the ROCm distributed timeout in `test/distributed/test_aten_comm_compute_reordering.py::TestComputeCommReorderingMultiProc::test_sink_waits`; the new job is queued now, so CI needs another revisit after it completes.

**Comment by jansel** [comment]
@claude are those CI failures related?

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main after the mergebot cherry-pick conflict in `test/inductor/test_codecache.py`, preserving the upstream `_find_triton_kernel_binaries` helper and the CPU thread-count cache-key tests, then resubmitted the stack. Validation passed with `PYTHONPATH=. python test/inductor/test_codecache.py -k test_hash_kwargs`, `PYTHONPATH=. python test/inductor/test_codecache.py -k test_cpu_thread_count`, `lintrunner -a`, and an independent review found no issues.

**Comment by jansel** [comment]
@claude are those CI failures related?

**Comment by jansel** [comment]
My agent says I inspected the current Dr. CI failures on head a745bcc14363e1bd0b88e7dd4c201480c2cea65f. The new/unclassified failures are in unrelated jobs/tests (`test_cpu_repro.py::CPUReproTests::test_nn_fold`, ROCm `test_inductor_freezing.py::FreezingCpuTests::test_as_strided_input_layout_with_symbolic_strides_cpu`, and torchtitan integration jobs) rather than the touched FX cache-key test/code paths, so I reran the failed pull, trunk, inductor, and torchtitan workflow jobs and will revisit after CI completes.

**Comment by jansel** [comment]
My agent says I rebased the PR past the autorevert and resubmitted it with a targeted follow-up for the  failure. The cache-key fix now makes the default-thread fold case compile a real parallel CPU kernel instead of reusing the one-thread cached kernel; because fold accumulates overlapping patches, that can differ from eager's serial accumulation order by small fp32 roundoff. I relaxed tolerance only for that default-thread branch. Validation passed for , , , ok No lint issues.
Successfully applied all patches., and diff whitespace checks.

**Comment by jansel** [comment]
My agent says correction to the previous comment: the validation commands that passed were `PYTHONPATH=. python test/inductor/test_codecache.py -k test_hash_kwargs`, `rm -rf /tmp/torchinductor_jansel/precompiled_headers && PYTHONPATH=. python test/inductor/test_codecache.py -k test_cpu_thread_count`, `PYTHONPATH=/tmp/codex_fake_torchvision:. python test/inductor/test_cpu_repro.py -k test_nn_fold`, `lintrunner -a`, and diff whitespace checks. The temporary torchvision shim was only to avoid this local environment having torchvision built for CUDA 13 while the rebuilt PyTorch is CUDA 12.8.
